driver-async/src/main/com/mongodb/async/client/ChangeStreamIterableImpl.java

```diff
@@ -0,0 +1,167 @@
+                             final Class<TResult> resultClass) {
+        this.namespace = notNull("namespace", namespace);
+        this.codecRegistry = notNull("codecRegistry", codecRegistry);
+        this.readPreference = notNull("readPreference", readPreference);
+        this.readConcern = notNull("readConcern", readConcern);
+        this.executor = notNull("executor", executor);
+        this.pipeline = notNull("pipeline", pipeline);
+        this.resultClass = notNull("resultClass", resultClass);
+    }
+
+    @Override
+    public ChangeStreamIterable<TResult> fullDocument(final FullDocument fullDocument) {
+        this.fullDocument = notNull("fullDocument", fullDocument);
+        return this;
+    }
+
+    @Override
+    public ChangeStreamIterable<TResult> resumeAfter(final Bson resumeAfter) {
+        this.resumeToken = notNull("resumeAfter", resumeAfter);
+        return this;
+    }
+
+    @Override
+    public ChangeStreamIterable<TResult> batchSize(final int batchSize) {
+        this.batchSize = notNull("batchSize", batchSize);
+        return this;
+    }
+
+    @Override
+    public ChangeStreamIterable<TResult> maxTime(final long maxTime, final TimeUnit timeUnit) {
+        notNull("timeUnit", timeUnit);
+        this.maxTimeMS = MILLISECONDS.convert(maxTime, timeUnit);
+        return this;
+    }
+
+    @Override
+    public ChangeStreamIterable<TResult> maxAwaitTime(final long maxAwaitTime, final TimeUnit timeUnit) {
+        notNull("timeUnit", timeUnit);
+        this.maxAwaitTimeMS = MILLISECONDS.convert(maxAwaitTime, timeUnit);
+        return this;
+    }
+
+    @Override
+    public ChangeStreamIterable<TResult> collation(final Collation collation) {
+        this.collation = notNull("collation", collation);
+        return this;
+    }
+
+    @Override
+    public void first(final SingleResultCallback<TResult> callback) {
+        notNull("callback", callback);
+        execute().first(callback);
+    }
+
+    @Override
+    public void forEach(final Block<? super TResult> block, final SingleResultCallback<Void> callback) {
+        notNull("block", block);
+        notNull("callback", callback);
+        execute().forEach(block, callback);
+    }
+
+    @Override
+    public <A extends Collection<? super TResult>> void into(final A target, final SingleResultCallback<A> callback) {
+        notNull("target", target);
+        notNull("callback", callback);
+        execute().into(target, callback);
+    }
+
+    @Override
+    public <U> MongoIterable<U> map(final Function<TResult, U> mapper) {
+        return new MappingIterable<TResult, U>(this, mapper);
+    }
+
+    @Override
+    public void batchCursor(final SingleResultCallback<AsyncBatchCursor<TResult>> callback) {
+        notNull("callback", callback);
+        execute().batchCursor(callback);
+    }
+
+    private MongoIterable<TResult> execute() {
```

Comment:
I'm surprised an Iterable has an execute method.

I would have expected an Iterable to be the *result* of calling execute on an operation.